Index: gcc/toplev.c
===================================================================
--- gcc/toplev.c.orig	2009-09-07 19:11:47.000000000 -0300
+++ gcc/toplev.c	2009-09-07 19:17:31.000000000 -0300
@@ -1796,7 +1796,12 @@ process_options (void)
   if (flag_gtoggle)
     {
       if (debug_info_level == DINFO_LEVEL_NONE)
-	debug_info_level = DINFO_LEVEL_NORMAL;
+	{
+	  debug_info_level = DINFO_LEVEL_NORMAL;
+
+	  if (write_symbols == NO_DEBUG)
+	    write_symbols = PREFERRED_DEBUGGING_TYPE;
+	}
       else
 	debug_info_level = DINFO_LEVEL_NONE;
     }
