--- libjava/testsuite/Makefile.am	2007-02-07 13:07:16.000000000 +0100
+++ libjava/testsuite/Makefile.am	2007-12-08 10:58:02.000000000 +0100
@@ -43,7 +43,7 @@ if JAVA_MAINTAINER_MODE
 	      testdep="libjava.loader/MyLoader.java";; \
 	    libjava.jni/register2.java) \
 	      ;; \
-	    libjava.jni/*.java) \
+	    libjava.jni/*.java|libjava.jvmti/interp/*.java) \
 	      genheader=-jni;; \
 	    libjava.jvmti/*.java|libjava.cni/*.java) \
 	      genheader=-cni;; \
--- libjava/testsuite/Makefile.in	2007-09-04 23:09:37.000000000 +0200
+++ libjava/testsuite/Makefile.in	2007-12-08 10:58:25.000000000 +0100
@@ -558,7 +558,7 @@ compile-tests:
 @JAVA_MAINTAINER_MODE_TRUE@	      testdep="libjava.loader/MyLoader.java";; \
 @JAVA_MAINTAINER_MODE_TRUE@	    libjava.jni/register2.java) \
 @JAVA_MAINTAINER_MODE_TRUE@	      ;; \
-@JAVA_MAINTAINER_MODE_TRUE@	    libjava.jni/*.java) \
+@JAVA_MAINTAINER_MODE_TRUE@	    libjava.jni/*.java|libjava.jvmti/interp/*.java) \
 @JAVA_MAINTAINER_MODE_TRUE@	      genheader=-jni;; \
 @JAVA_MAINTAINER_MODE_TRUE@	    libjava.jvmti/*.java|libjava.cni/*.java) \
 @JAVA_MAINTAINER_MODE_TRUE@	      genheader=-cni;; \
