--- libjava/classpath/configure.ac	(revision 130691)
+++ libjava/classpath/configure.ac	(working copy)
@@ -294,14 +294,14 @@ CLASSPATH_TOOLEXECLIBDIR
 dnl -----------------------------------------------------------
 dnl Sets the native libraries installation dir
 dnl -----------------------------------------------------------
-dnl GCJ LOCAL: default to ${toolexeclibdir}/gcj-${gcc_version}-${libgcj_soversion}
+dnl GCJ LOCAL: default to ${toolexeclibdir}/gcj-${gcc_version}
 AC_ARG_WITH([native-libdir],
 	    [AS_HELP_STRING(--with-native-libdir,sets the installation directory for native libraries [default='${libdir}/${PACKAGE}'])],
 	    [
 	     nativeexeclibdir=${withval}
 	    ],
 	    [
-	     nativeexeclibdir='${toolexeclibdir}/gcj-'`cat ${srcdir}/../../gcc/BASE-VER`-`awk -F: '/^[[^#]].*:/ { print $1 }' ${srcdir}/../libtool-version`
+	     nativeexeclibdir='${toolexeclibdir}/gcj-'`cat ${srcdir}/../../gcc/BASE-VER`
 	    ])
 
 AC_SUBST(nativeexeclibdir)
--- libjava/classpath/configure	(revision 130691)
+++ libjava/classpath/configure	(working copy)
@@ -4771,7 +4771,7 @@ if test "${with_native_libdir+set}" = se
 
 else
 
-	     nativeexeclibdir='${toolexeclibdir}/gcj-'`cat ${srcdir}/../../gcc/BASE-VER`-`awk -F: '/^[^#].*:/ { print $1 }' ${srcdir}/../libtool-version`
+	     nativeexeclibdir='${toolexeclibdir}/gcj-'`cat ${srcdir}/../../gcc/BASE-VER`
 
 fi;
 
