diff -up ./libjava/configure.ac.sav ./libjava/configure.ac
--- ./libjava/configure.ac.sav	2009-04-03 14:56:35.000000000 -0400
+++ ./libjava/configure.ac	2009-04-03 14:56:50.000000000 -0400
@@ -1881,7 +1881,7 @@ then
     case ${host} in
       *-mingw* | *-cygwin*)
         host_cpu=x86;;
-      i486-* | i586-* | i686-*)
+      i386-* | i486-* | i586-* | i686-*)
         host_cpu=i386;;
       x86_64-*)
         host_cpu=amd64;;
diff -up ./libjava/configure.sav ./libjava/configure
--- ./libjava/configure.sav	2009-04-03 14:56:27.000000000 -0400
+++ ./libjava/configure	2009-04-03 15:02:55.000000000 -0400
@@ -28285,7 +28285,7 @@ echo "${ECHO_T}host is ${host}" >&6
     case ${host} in
       *-mingw* | *-cygwin*)
         host_cpu=x86;;
-      i486-* | i586-* | i686-*)
+      i386-* | i486-* | i586-* | i686-*)
         host_cpu=i386;;
       x86_64-*)
         host_cpu=amd64;;
