# Workaround doxygen 1.8.10 bugs.
--- libstdc++-v3/doc/doxygen/user.cfg.in
+++ libstdc++-v3/doc/doxygen/user.cfg.in
@@ -895,7 +895,6 @@ INPUT                  = @srcdir@/doc/doxygen/doxygroups.cc \
                          include/ext/pb_ds/detail/binary_heap_ \
                          include/ext/pb_ds/detail/binomial_heap_ \
                          include/ext/pb_ds/detail/binomial_heap_base_ \
-                         include/ext/pb_ds/detail/bin_search_tree_ \
                          include/ext/pb_ds/detail/branch_policy \
                          include/ext/pb_ds/detail/cc_hash_table_map_ \
                          include/ext/pb_ds/detail/eq_fn \
@@ -1965,18 +1964,6 @@ GENERATE_XML           = @do_xml@
 
 XML_OUTPUT             = xml
 
-# The XML_SCHEMA tag can be used to specify a XML schema, which can be used by a
-# validating XML parser to check the syntax of the XML files.
-# This tag requires that the tag GENERATE_XML is set to YES.
-
-XML_SCHEMA             =
-
-# The XML_DTD tag can be used to specify a XML DTD, which can be used by a
-# validating XML parser to check the syntax of the XML files.
-# This tag requires that the tag GENERATE_XML is set to YES.
-
-XML_DTD                =
-
 # If the XML_PROGRAMLISTING tag is set to YES doxygen will dump the program
 # listings (including syntax highlighting and cross-referencing information) to
 # the XML output. Note that enabling this will significantly increase the size
@@ -2135,6 +2122,7 @@ PREDEFINED             = __cplusplus=201103L \
                          _GLIBCXX_USE_CONSTEXPR=constexpr \
                          "_GLIBCXX_THROW(E)= " \
                          _GLIBCXX_NOEXCEPT=noexcept \
+                         "_GLIBCXX_NOEXCEPT_IF(E)=noexcept(E)" \
                          _GLIBCXX_NOTHROW=noexcept \
                          _GLIBCXX_USE_NOEXCEPT=noexcept \
                          _GLIBCXX_USE_WCHAR_T \
